File: nginx_test/was-4-springboot/src/main/java/com/nginx/test/controller/ApiController.java
package com.nginx.test.controller;

import com.nginx.test.service.CacheService;
import com.nginx.test.service.MessageService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.net.InetAddress;
import java.time.Instant;
import java.util.HashMap;
import java.util.Map;

@Slf4j
@RestController
@RequestMapping("/api")
@RequiredArgsConstructor
public class ApiController {

    private final CacheService cacheService;
    private final MessageService messageService;

    @GetMapping("/info")
    public ResponseEntity<Map<String, Object>> getInfo() {
        log.info("Received info request");

        Map<String, Object> info = new HashMap<>();
        info.put("service", "was-4-springboot");
        info.put("type", "kubernetes-docker-runtime");
        info.put("framework", "Spring Boot 3.2 + Java 21");
        info.put("timestamp", Instant.now().toString());

        try {
            info.put("hostname", InetAddress.getLocalHost().getHostName());
            info.put("ip", InetAddress.getLocalHost().getHostAddress());
        } catch (Exception e) {
            info.put("hostname", "unknown");
            info.put("ip", "unknown");
        }

        info.put("javaVersion", System.getProperty("java.version"));

        Map<String, String> environment = new HashMap<>();
        environment.put("POD_NAME", System.getenv().getOrDefault("POD_NAME", "local"));
        environment.put("POD_NAMESPACE", System.getenv().getOrDefault("POD_NAMESPACE", "default"));
        environment.put("NODE_NAME", System.getenv().getOrDefault("NODE_NAME", "local"));
        info.put("environment", environment);

        return ResponseEntity.ok(info);
    }

    @GetMapping("/health")
    public ResponseEntity<Map<String, Object>> healthCheck() {
        Map<String, Object> health = new HashMap<>();
        health.put("status", "UP");
        health.put("service", "was-4-springboot");
        health.put("timestamp", Instant.now().toString());

        Map<String, Object> checks = new HashMap<>();
        checks.put("redis", cacheService.healthCheck());
        checks.put("kafka", messageService.healthCheck());
        health.put("checks", checks);

        return ResponseEntity.ok(health);
    }

    @GetMapping("/cache/{key}")
    public ResponseEntity<Map<String, Object>> getCache(@PathVariable String key) {
        log.info("Cache GET request for key: {}", key);

        String value = cacheService.get(key);

        Map<String, Object> result = new HashMap<>();
        result.put("operation", "GET");
        result.put("key", key);
        result.put("value", value);
        result.put("found", value != null);
        result.put("source", "redis-cluster");

        return ResponseEntity.ok(result);
    }

    @PutMapping("/cache/{key}")
    public ResponseEntity<Map<String, Object>> setCache(
            @PathVariable String key,
            @RequestBody Map<String, Object> body) {
        log.info("Cache SET request for key: {}", key);

        String value = body.getOrDefault("value", "").toString();
        long ttl = body.containsKey("ttl") ? ((Number) body.get("ttl")).longValue() : 3600L;

        cacheService.set(key, value, ttl);

        Map<String, Object> result = new HashMap<>();
        result.put("operation", "SET");
        result.put("key", key);
        result.put("value", value);
        result.put("ttl", ttl);
        result.put("success", true);
        result.put("destination", "redis-cluster");

        return ResponseEntity.ok(result);
    }

    @DeleteMapping("/cache/{key}")
    public ResponseEntity<Map<String, Object>> deleteCache(@PathVariable String key) {
        log.info("Cache DELETE request for key: {}", key);

        boolean deleted = cacheService.delete(key);

        Map<String, Object> result = new HashMap<>();
        result.put("operation", "DELETE");
        result.put("key", key);
        result.put("deleted", deleted);
        result.put("destination", "redis-cluster");

        return ResponseEntity.ok(result);
    }

    @PostMapping("/message")
    public ResponseEntity<Map<String, Object>> sendMessage(@RequestBody Map<String, Object> body) {
        String topic = body.getOrDefault("topic", "nginx-test-events").toString();
        String message = body.getOrDefault("message", "").toString();
        String key = body.containsKey("key") ? body.get("key").toString() : null;

        log.info("Sending message to Kafka topic: {}", topic);

        String messageId = messageService.send(topic, key, message);

        Map<String, Object> result = new HashMap<>();
        result.put("operation", "PUBLISH");
        result.put("topic", topic);
        result.put("key", key != null ? key : "null");
        result.put("messageId", messageId);
        result.put("success", true);
        result.put("broker", "kafka-cluster");

        return ResponseEntity.ok(result);
    }

    @GetMapping("/slow")
    public ResponseEntity<Map<String, Object>> slowEndpoint(
            @RequestParam(defaultValue = "1000") long delay) throws InterruptedException {
        log.info("Slow endpoint called with delay: {}ms", delay);

        Thread.sleep(delay);

        Map<String, Object> result = new HashMap<>();
        result.put("service", "was-4-springboot");
        result.put("endpoint", "/api/slow");
        result.put("delay_ms", delay);
        result.put("message", "This endpoint simulates slow responses for testing timeouts");

        return ResponseEntity.ok(result);
    }

    @GetMapping("/error")
    public ResponseEntity<Map<String, Object>> errorEndpoint(
            @RequestParam(defaultValue = "500") int code) {
        log.error("Error endpoint called with code: {}", code);

        Map<String, Object> result = new HashMap<>();
        result.put("service", "was-4-springboot");
        result.put("endpoint", "/api/error");
        result.put("error_code", code);
        result.put("message", "This endpoint simulates errors for testing error handling");

        return ResponseEntity.status(code).body(result);
    }
}

File: nginx_test/was-3-springboot/src/main/kotlin/com/nginx/test/controller/ApiController.kt
package com.nginx.test.controller

import com.nginx.test.service.CacheService
import com.nginx.test.service.MessageService
import io.micrometer.observation.annotation.Observed
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.net.InetAddress
import java.time.Instant

@RestController
@RequestMapping("/api")
@Observed(name = "api.controller")
class ApiController(
    private val cacheService: CacheService,
    private val messageService: MessageService
) {
    private val logger = LoggerFactory.getLogger(javaClass)

    @GetMapping("/info")
    fun getInfo(): ResponseEntity<Map<String, Any>> {
        logger.info("Received info request")

        val info = mapOf(
            "service" to "was-3-springboot",
            "type" to "kubernetes-docker-runtime",
            "framework" to "Spring Boot 3.2 + Kotlin",
            "timestamp" to Instant.now().toString(),
            "hostname" to try { InetAddress.getLocalHost().hostName } catch (e: Exception) { "unknown" },
            "ip" to try { InetAddress.getLocalHost().hostAddress } catch (e: Exception) { "unknown" },
            "javaVersion" to System.getProperty("java.version"),
            "environment" to mapOf(
                "POD_NAME" to (System.getenv("POD_NAME") ?: "local"),
                "POD_NAMESPACE" to (System.getenv("POD_NAMESPACE") ?: "default"),
                "NODE_NAME" to (System.getenv("NODE_NAME") ?: "local")
            )
        )

        return ResponseEntity.ok(info)
    }

    @GetMapping("/health")
    fun healthCheck(): ResponseEntity<Map<String, Any>> {
        val health = mapOf(
            "status" to "UP",
            "service" to "was-3-springboot",
            "timestamp" to Instant.now().toString(),
            "checks" to mapOf(
                "redis" to cacheService.healthCheck(),
                "kafka" to messageService.healthCheck()
            )
        )
        return ResponseEntity.ok(health)
    }

    @GetMapping("/cache/{key}")
    fun getCache(@PathVariable key: String): ResponseEntity<Map<String, Any?>> {
        logger.info("Cache GET request for key: $key")

        val value = cacheService.get(key)
        val result = mapOf(
            "operation" to "GET",
            "key" to key,
            "value" to value,
            "found" to (value != null),
            "source" to "redis-cluster"
        )

        return ResponseEntity.ok(result)
    }

    @PutMapping("/cache/{key}")
    fun setCache(
        @PathVariable key: String,
        @RequestBody body: Map<String, Any>
    ): ResponseEntity<Map<String, Any>> {
        logger.info("Cache SET request for key: $key")

        val value = body["value"]?.toString() ?: ""
        val ttl = (body["ttl"] as? Number)?.toLong() ?: 3600L

        cacheService.set(key, value, ttl)

        val result = mapOf(
            "operation" to "SET",
            "key" to key,
            "value" to value,
            "ttl" to ttl,
            "success" to true,
            "destination" to "redis-cluster"
        )

        return ResponseEntity.ok(result)
    }

    @DeleteMapping("/cache/{key}")
    fun deleteCache(@PathVariable key: String): ResponseEntity<Map<String, Any>> {
        logger.info("Cache DELETE request for key: $key")

        val deleted = cacheService.delete(key)

        val result = mapOf(
            "operation" to "DELETE",
            "key" to key,
            "deleted" to deleted,
            "destination" to "redis-cluster"
        )

        return ResponseEntity.ok(result)
    }

    @PostMapping("/message")
    fun sendMessage(@RequestBody body: Map<String, Any>): ResponseEntity<Map<String, Any>> {
        val topic = body["topic"]?.toString() ?: "nginx-test-events"
        val message = body["message"]?.toString() ?: ""
        val key = body["key"]?.toString()

        logger.info("Sending message to Kafka topic: $topic")

        val messageId = messageService.send(topic, key, message)

        val result = mapOf(
            "operation" to "PUBLISH",
            "topic" to topic,
            "key" to (key ?: "null"),
            "messageId" to messageId,
            "success" to true,
            "broker" to "kafka-cluster"
        )

        return ResponseEntity.ok(result)
    }

    @GetMapping("/slow")
    fun slowEndpoint(@RequestParam(defaultValue = "1000") delay: Long): ResponseEntity<Map<String, Any>> {
        logger.info("Slow endpoint called with delay: ${delay}ms")

        Thread.sleep(delay)

        val result = mapOf(
            "service" to "was-3-springboot",
            "endpoint" to "/api/slow",
            "delay_ms" to delay,
            "message" to "This endpoint simulates slow responses for testing timeouts"
        )

        return ResponseEntity.ok(result)
    }

    @GetMapping("/error")
    fun errorEndpoint(@RequestParam(defaultValue = "500") code: Int): ResponseEntity<Map<String, Any>> {
        logger.error("Error endpoint called with code: $code")

        val result = mapOf(
            "service" to "was-3-springboot",
            "endpoint" to "/api/error",
            "error_code" to code,
            "message" to "This endpoint simulates errors for testing error handling"
        )

        return ResponseEntity.status(code).body(result)
    }
}
